// material-dynamic-table/esm5/lib/table-cell/cell-types/date-cell.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ColumnConfig } from '../../column-config.model';
var DateCellComponent = /** @class */ (function () {
    function DateCellComponent() {
        this.dateFormat = 'short';
    }
    /**
     * @return {?}
     */
    DateCellComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.column.options) {
            if (this.column.options.dateFormat) {
                this.dateFormat = this.column.options.dateFormat;
            }
        }
    };
    DateCellComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mdt-date-cell',
                    template: '{{ row[column.name] | date:dateFormat }}'
                }] }
    ];
    DateCellComponent.propDecorators = {
        column: [{ type: Input }],
        row: [{ type: Input }]
    };
    return DateCellComponent;
}());
export { DateCellComponent };
if (false) {
    /** @type {?} */
    DateCellComponent.prototype.column;
    /** @type {?} */
    DateCellComponent.prototype.row;
    /** @type {?} */
    DateCellComponent.prototype.dateFormat;
}
