// material-dynamic-table/esm5/lib/table-cell/table-cell.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ComponentFactoryResolver, Input, ViewChild } from '@angular/core';
import { CellDirective } from './cell.directive';
import { CellService } from './cell-types/cell.service';
import { ColumnConfig } from '../column-config.model';
var TableCellComponent = /** @class */ (function () {
    function TableCellComponent(cellService, componentFactoryResolver) {
        this.cellService = cellService;
        this.componentFactoryResolver = componentFactoryResolver;
    }
    /**
     * @return {?}
     */
    TableCellComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initCell();
    };
    /**
     * @return {?}
     */
    TableCellComponent.prototype.initCell = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var cellComponent = this.cellService.getCell(this.column.type);
        /** @type {?} */
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(cellComponent);
        /** @type {?} */
        var viewContainerRef = this.cellHost.viewContainerRef;
        viewContainerRef.clear();
        /** @type {?} */
        var componentRef = viewContainerRef.createComponent(componentFactory);
        /** @type {?} */
        var cell = (/** @type {?} */ (componentRef.instance));
        cell.row = this.row;
        cell.column = this.column;
    };
    TableCellComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mdt-table-cell',
                    template: '<ng-template mdtCellHost></ng-template>'
                }] }
    ];
    /** @nocollapse */
    TableCellComponent.ctorParameters = function () { return [
        { type: CellService },
        { type: ComponentFactoryResolver }
    ]; };
    TableCellComponent.propDecorators = {
        cellHost: [{ type: ViewChild, args: [CellDirective,] }],
        row: [{ type: Input }],
        column: [{ type: Input }]
    };
    return TableCellComponent;
}());
export { TableCellComponent };
if (false) {
    /** @type {?} */
    TableCellComponent.prototype.cellHost;
    /** @type {?} */
    TableCellComponent.prototype.row;
    /** @type {?} */
    TableCellComponent.prototype.column;
    /**
     * @type {?}
     * @private
     */
    TableCellComponent.prototype.cellService;
    /**
     * @type {?}
     * @private
     */
    TableCellComponent.prototype.componentFactoryResolver;
}
