// material-dynamic-table/esm2015/lib/table-cell/cell-types/date-cell.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ColumnConfig } from '../../column-config.model';
export class DateCellComponent {
    constructor() {
        this.dateFormat = 'short';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.column.options) {
            if (this.column.options.dateFormat) {
                this.dateFormat = this.column.options.dateFormat;
            }
        }
    }
}
DateCellComponent.decorators = [
    { type: Component, args: [{
                selector: 'mdt-date-cell',
                template: '{{ row[column.name] | date:dateFormat }}'
            }] }
];
DateCellComponent.propDecorators = {
    column: [{ type: Input }],
    row: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DateCellComponent.prototype.column;
    /** @type {?} */
    DateCellComponent.prototype.row;
    /** @type {?} */
    DateCellComponent.prototype.dateFormat;
}
