// material-dynamic-table/esm2015/lib/table-cell/table-cell.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ComponentFactoryResolver, Input, ViewChild } from '@angular/core';
import { CellDirective } from './cell.directive';
import { CellService } from './cell-types/cell.service';
import { ColumnConfig } from '../column-config.model';
export class TableCellComponent {
    /**
     * @param {?} cellService
     * @param {?} componentFactoryResolver
     */
    constructor(cellService, componentFactoryResolver) {
        this.cellService = cellService;
        this.componentFactoryResolver = componentFactoryResolver;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.initCell();
    }
    /**
     * @return {?}
     */
    initCell() {
        /** @type {?} */
        const cellComponent = this.cellService.getCell(this.column.type);
        /** @type {?} */
        const componentFactory = this.componentFactoryResolver.resolveComponentFactory(cellComponent);
        /** @type {?} */
        const viewContainerRef = this.cellHost.viewContainerRef;
        viewContainerRef.clear();
        /** @type {?} */
        const componentRef = viewContainerRef.createComponent(componentFactory);
        /** @type {?} */
        const cell = (/** @type {?} */ (componentRef.instance));
        cell.row = this.row;
        cell.column = this.column;
    }
}
TableCellComponent.decorators = [
    { type: Component, args: [{
                selector: 'mdt-table-cell',
                template: '<ng-template mdtCellHost></ng-template>'
            }] }
];
/** @nocollapse */
TableCellComponent.ctorParameters = () => [
    { type: CellService },
    { type: ComponentFactoryResolver }
];
TableCellComponent.propDecorators = {
    cellHost: [{ type: ViewChild, args: [CellDirective,] }],
    row: [{ type: Input }],
    column: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TableCellComponent.prototype.cellHost;
    /** @type {?} */
    TableCellComponent.prototype.row;
    /** @type {?} */
    TableCellComponent.prototype.column;
    /**
     * @type {?}
     * @private
     */
    TableCellComponent.prototype.cellService;
    /**
     * @type {?}
     * @private
     */
    TableCellComponent.prototype.componentFactoryResolver;
}
