// material-dynamic-table/esm2015/lib/dynamic-table.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatTableModule, MatSortModule, MatPaginatorModule, MatIconModule, MatDialogModule, MatTooltipModule } from '@angular/material';
import { DynamicTableComponent } from './dynamic-table.component';
import { TableCellComponent } from './table-cell/table-cell.component';
import { CellService } from './table-cell/cell-types/cell.service';
import { CellDirective } from './table-cell/cell.directive';
import { ColumnFilterService } from './table-cell/cell-types/column-filter.service';
export { CellService, CellDirective, ColumnFilterService };
export { ColumnFilter } from './column-filter.model';
export { ColumnConfig } from './column-config.model';
import { TextCellComponent } from './table-cell/cell-types/text-cell.component';
import { DateCellComponent } from './table-cell/cell-types/date-cell.component';
export class DynamicTableModule {
    /**
     * @param {?} cellService
     */
    constructor(cellService) {
        this.cellService = cellService;
        cellService.registerCell('string', TextCellComponent);
        cellService.registerCell('date', DateCellComponent);
    }
}
DynamicTableModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    MatTableModule,
                    MatSortModule,
                    MatPaginatorModule,
                    MatIconModule,
                    MatDialogModule,
                    MatTooltipModule
                ],
                declarations: [
                    DynamicTableComponent,
                    TableCellComponent,
                    CellDirective,
                    TextCellComponent,
                    DateCellComponent
                ],
                exports: [DynamicTableComponent],
                entryComponents: [
                    TextCellComponent,
                    DateCellComponent
                ],
                providers: [
                    CellService,
                    ColumnFilterService
                ]
            },] }
];
/** @nocollapse */
DynamicTableModule.ctorParameters = () => [
    { type: CellService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    DynamicTableModule.prototype.cellService;
}
